### PR TITLE
Add warnings to on_before/after_batch_transfer hooks

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Adds warnings that the API will change in the future to support dataloader idx.

Related to #6058 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by edenlightning** [comment]
For other features we just say “This hook is in beta and is subject to change", but yours is good too.

**Review by edenlightning — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Comment by rohitgr7** [comment]
@SeanNaren add the same for `transfer_batch_to_device` hook too?

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 1]
```diff
@@ -305,7 +305,13 @@ def _format_precision_config(self):
...
         if precision == 16:
             if "amp" not in self.config and amp_type == AMPType.NATIVE:
-                self.config["fp16"] = {"enabled": True}
+                self.config["fp16"] = {
+                    "enabled": True,
+                    "loss_scale": 0,
+                    "loss_scale_window": 1000,
+                    "hysteresis": 2,
+                    "min_loss_scale": 1
+                }
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:314]: sneaking in deepspeed changes huh!!!

Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:314]: 🤦

**Comment by SeanNaren** [comment]
@rohitgr7 we've modified it slightly after internal discussion, by adding a dummy dataloader_idx and a warning to the docstrings that this always returns 0. Shall we do the same for transfer_batch_to_device? cc @carmocca

**Review by justusschock — APPROVED** [review]

File: docs/source/extensions/datamodules.rst [hunk 2]
```diff
@@ -297,13 +297,16 @@ Override to define how you want to move an arbitrary batch to a device.
...
-        def transfer_batch_to_device(self, batch, device):
+        def transfer_batch_to_device(self, batch, device, dataloader_idx):
             x = batch['x']
             x = CustomDataWrapper(x)
             batch['x'] = x.to(device)
             return batch
 
 
+.. warning::
+    Currently dataloader_idx always returns 0 and will be updated to support the true idx in the future.
```
Comment (Borda) [inline docs/source/extensions/datamodules.rst:308]: not sure if there shall be a blank line in between...
```suggestion

    Currently ``dataloader_idx`` always returns 0 and will be updated to support the true ``idx`` in the future.
```

File: pytorch_lightning/core/hooks.py [hunk 3]
```diff
@@ -565,11 +565,13 @@ def predict_dataloader(self) -> Union[DataLoader, List[DataLoader]]:
             will have an argument ``dataloader_idx`` which matches the order here.
         """
 
-    def transfer_batch_to_device(self, batch: Any, device: Optional[torch.device] = None) -> Any:
+    def transfer_batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx=0) -> Any:
         """
         Override this hook if your :class:`~torch.utils.data.DataLoader` returns tensors
         wrapped in a custom data structure.
 
+        .. warning:: dataloader_idx always returns 0, and will be updated to support the true idx in the future.
```
Comment (Borda) [inline pytorch_lightning/core/hooks.py:573]:
```suggestion
        .. warning:: ``dataloader_idx`` always returns 0, and will be updated to support the true ``idx`` in the future.
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/core/hooks.py [hunk 4]
```diff
@@ -594,19 +596,20 @@ def transfer_batch_to_device(self, batch: Any, device: Optional[torch.device] =
         Args:
             batch: A batch of data that needs to be transferred to a new device.
             device: The target device as defined in PyTorch.
+            dataloader_idx: DataLoader idx for batch (Default: 0)
```
Comment (Borda) [inline pytorch_lightning/core/hooks.py:599]:
```suggestion
            dataloader_idx: DataLoader idx for batch
```

File: pytorch_lightning/core/hooks.py [hunk 5]
```diff
@@ -616,22 +619,25 @@ def transfer_batch_to_device(self, batch, device):
...
         Override to alter or apply batch augmentations to your batch before it is transferred to the device.
 
+        .. warning:: dataloader_idx always returns 0, and will be updated to support the true idx in the future.
+
         Note:
             This hook only runs on single GPU training and DDP (no data-parallel).
 
         Args:
             batch: A batch of data that needs to be altered or augmented.
+            dataloader_idx: DataLoader idx for batch (Default: 0)
```
Comment (Borda) [inline pytorch_lightning/core/hooks.py:633]:
```suggestion
            dataloader_idx: DataLoader idx for batch
```

File: pytorch_lightning/core/hooks.py [hunk 6]
```diff
@@ -641,22 +647,25 @@ def on_before_batch_transfer(self, batch):
         """
         return batch
 
-    def on_after_batch_transfer(self, batch):
+    def on_after_batch_transfer(self, batch, dataloader_idx):
         """
         Override to alter or apply batch augmentations to your batch after it is transferred to the device.
 
+        .. warning:: dataloader_idx always returns 0, and will be updated to support the true idx in the future.
```
Comment (Borda) [inline pytorch_lightning/core/hooks.py:654]:
```suggestion
        .. warning:: ``dataloader_idx`` always returns 0, and will be updated to support the true ``idx`` in the future.
```

File: docs/source/extensions/datamodules.rst [hunk 7]
```diff
@@ -329,11 +335,14 @@ Override to alter or apply augmentations to your batch after it is transferred t
...
 
     class MNISTDataModule(LightningDataModule):
-        def on_after_batch_transfer(self, batch):
+        def on_after_batch_transfer(self, batch, dataloader_idx):
             batch['x'] = gpu_transforms(batch['x'])
             return batch
 
 
+.. warning::
+    Currently dataloader_idx always returns 0 and will be updated to support the true idx in the future.
```
Comment (Borda) [inline docs/source/extensions/datamodules.rst:344]:
```suggestion

    Currently ``dataloader_idx`` always returns 0 and will be updated to support the true ``idx`` in the future.
```

**Review by rohitgr7 — APPROVED** [review]

**Comment by SeanNaren** [comment]
I'm questioning whether we should change `transfer_batch_to_device`, since it's already being used in the API for a while... any thoughts?

**Comment by edenlightning** [comment]
mmm if it'll break peoples code lets not change that

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by rohitgr7** [comment]
> I'm questioning whether we should change `transfer_batch_to_device`, since it's already being used in the API for a while... any thoughts?
> 
> @awaelchli brought up a good point in which if someone has overridden the `transfer_batch_to_device` hook already, then this will now throw an error right?

so the hooks that are already being there for a while won't be updated ever again :thinking: ? If someone's going to upgrade the package then only it will be a problem right. I thought the point of doing feature releases is to add new features and upgrade the old ones if required.
